Add timestamps to `pvm_send` abstraction and `pvm.py`

KaneBetter opened the pull request:
This fixes #35
This fixes #5

KaneBetter committed:
Add time in pvm_send.maxpat

KaneBetter committed:
Add time parse function in pvm.py

KaneBetter committed:
Revise pvm_send

omarcostahamido commented on pvm.py:
```diff
@@ -30,18 +31,40 @@ def _init_logger():
+	global NEXT_TIME
+	if len(args) == 4:
+		seconds = int(args[3]) + 3
+		mintues = int(args[2])
+		hours = int(args[1])
+		if seconds >= 60:
+			seconds -= 60
+			mintues += 1
+		if mintues == 60:
+			mintues = 0
+			hours += 1
+		time_data = str(hours) + ":" + str(mintues) + ":" + str(seconds)
```
please fix `mintues` > `minutes`

KaneBetter replied:
I will test the new MAX file again.
For the previous version of pvm_send. For each rpi, MAX will send the timestamp first and then send the command.

omarcostahamido replied:
that's what I made sure it would still happen: first send timestamp and then send command.
As it was, it was sending timestamp even when it was not sending a command, for example, when the `pvm_send` abstraction received messages to change its state.

omarcostahamido commented on pvm.py:
```diff
@@ -30,18 +31,40 @@ def _init_logger():
+		_logger.info("Received value: %s", str(value))
 		pass
 	# TODO: Create another python file to control two display
+	while True:
+		now = datetime.now()
+		now = now.replace(microsecond=0)
+		if now.time() == NEXT_TIME.time():
+			break
+		sleep(0.01)
```
interesting... does that mean that the rpi will become unresponsive for 3 seconds?
also, it seems that if for some reason the clock of the rpi device is not in sync with the control computer then chaos will be easily instaured: e.g. if the timestamp is for 18:00:00 (to which this script will add 3 seconds) and the rpi is at 18:00:04, it means that the target time will never reach and the rpi device will become permanently unresponsive until it is manually reset?

KaneBetter replied:
I haven't tested the situation in RPIs where press start and pause in the 3s.
But from the logic of the code, it will be permanently unresponsive. Fix this in the newest commit.

omarcostahamido reviewed (commented):
some comments on the python script only. will check out the patch now.
Ok, notes on the revision of the patch:
- I will try to add a simple commit to clear unnecessary things
- you don't want to trigger the sending of timestamps with every new message that gets to `pvm_send` abstraction, at most with every message that gets sent to `udp` object inside it, which is a smaller subsection.
- can you confirm that this timestamp will be needed for every single successful command that is sent to the rpi devices?

omarcostahamido committed:
Update pvm_send.maxpat

- avoid unnecessary triggering of `date` object
- ensure proper order of events: first send timestamp, then the command
- avoid passing _done_ message from `patcherargs` object

omarcostahamido commented:
@KaneBetter please notice the commit I've added here.

KaneBetter commented:
> @KaneBetter please notice the commit I've added here.

Thank you for this commit. I'm really bad at MSP.

KaneBetter committed:
Fix permanently unresponsive bug for multiple actions in 3s

omarcostahamido commented:
> Thank you for this commit. I'm really bad at MSP.

ha. no worries. I'll continue to oversee that part.

KaneBetter committed:
Return NEXT_TIME <= time.now()

omarcostahamido reviewed (approved):
@KaneBetter thx for the updates.
I am approving this now, so that you can merge when/if you want. 
However, I really feel that this needs to be tested in context. It feels that waiting 3 seconds for any action to take place will seriously hinder the possibilities of using PVM as a realtime interactive tool (imagine sort of a _VJ-ing_ context), which would be the end goal.
@gggfffsss please chime in and give your informed view on this too when you can.

KaneBetter commented:
> @KaneBetter thx for the updates. I am approving this now, so that you can merge when/if you want. However, I really feel that this needs to be tested in context. It feels that waiting 3 seconds for any action to take place will seriously hinder the possibilities of using PVM as a realtime interactive tool (imagine sort of a _VJ-ing_ context), which would be the end goal. @gggfffsss please chime in and give your informed view on this too when you can.

In the case of 2 Raspberry Pi we tested 1s latency is enough, the reason why we chose 3s is because we have not tested the environment with 6 Raspberry Pi, we guessed that the case of 6 Raspberry Pi may lead to 2s latency, so we chose 3s.

KaneBetter commented:
> @KaneBetter thx for the updates. I am approving this now, so that you can merge when/if you want. However, I really feel that this needs to be tested in context. It feels that waiting 3 seconds for any action to take place will seriously hinder the possibilities of using PVM as a realtime interactive tool (imagine sort of a _VJ-ing_ context), which would be the end goal. @gggfffsss please chime in and give your informed view on this too when you can.

Also I will push another commit to this branch so that it can ignore commands that are out of date(behind the current time).

KaneBetter committed:
Add comment and fix the command pack(2 commands) logic

omarcostahamido commented:
> In the case of 2 Raspberry Pi we tested 1s latency is enough, the reason why we chose 3s is because we have not tested the environment with 6 Raspberry Pi, we guessed that the case of 6 Raspberry Pi may lead to 2s latency, so we chose 3s.

Please perform the tests, you have access to the equipment and space!

KaneBetter committed:
Merge branch 'main' into AddTimeInUDP

omarcostahamido commented on pvm.py:
```diff
@@ -36,29 +37,90 @@ def _init_logger():
+
+'''
+UDP command example:
+/PVM 12:13:14
```
now that I am seeing this here I am realizing that the current way to define and parse the timestamps are really fragile. We need to add a `timestamp` tag, like all the other commands so here it should look like
```suggestion
/PVM timestamp HH MM SS
```

KaneBetter replied:
Is it possible to let MAX send a command like below
`/PVM HH MM SS <action_name> <action value>`
which will reduce a lot of logic complexity.

omarcostahamido replied:
sure thing. give me a sec.

KaneBetter replied:
Thank you

omarcostahamido commented on pvm.py:
```diff
@@ -36,29 +37,90 @@ def _init_logger():
+	global NEXT_TIME
+	# Parse the time send in UDP, 
+	# From /PVM 12 13 14 to 12:13:14
+	if len(args) == 4:
+		seconds = int(args[3]) + 3
+		minutes = int(args[2])
+		hours = int(args[1])
```
yup, I can definitely see this being very problematic here, if we have some other command that also has 3 values then it will be parsed here and the program may not like that surprize.
```suggestion
	if args[1]=="timestamp" && len(args) == 5:
		seconds = int(args[4]) + 3
		minutes = int(args[3])
		hours = int(args[2])
```

omarcostahamido replied:
following the new commit https://github.com/omarcostahamido/PVM/pull/38/commits/33404ff646efb2093db36b51a4a6df5158cb5c23 this would have to be different now

KaneBetter replied:
Got it. Will change and test ASAP.

omarcostahamido reviewed (changes requested):
I am adding a commit to change the `pvm_send` abstraction to send the `timestamp` tag, so that we can avoid the problems I reported below.

omarcostahamido committed:
Update pvm_send.maxpat

- add `timestamp` tag to `HH MM SS` message so that it is easier/less problematic to parse in `pvm.py`

omarcostahamido commented:
@KaneBetter please notice the new commit I've added here.

omarcostahamido committed:
combine timestamp & commands

- following request by @KaneBetter in https://github.com/omarcostahamido/PVM/pull/38#discussion_r1020951438

omarcostahamido commented:
@KaneBetter please notice the new commit I've added here in response to your request above.

KaneBetter commented:
After I pull the latest version of pvm.maxpat. I tried to test it and I found out that there is no udp commands sent out by MAX. And I tried the version on the main branch, it works fine. So my MAX software doesn't have a problem. The problem is in this version of pvm.maxpat. @omarcostahamido

omarcostahamido commented:
@KaneBetter let me take a look

omarcostahamido commented:
@KaneBetter can you confirm that if you just click the `reset` message on `pvm.maxpat` it then just works?
I think there is only a problem with the order of messages, I can try to fiddle around to force the order we need...

omarcostahamido committed:
`deferlow` the loadmess on `pvm_init`

- fixing the order of events when loading the `pvm.maxpat`

omarcostahamido commented:
@KaneBetter please note the new commit.
It is strange that it worked well before and not right now.
In any case this new addition to `pvm_init` should be able to enforce the correct order of events startup.

KaneBetter commented:
Unfortunately, the newest one also didn't work. When I hit the reset button, it shows below. 
![image](https://user-images.githubusercontent.com/37838032/201999563-ebf67c6f-f727-4a8a-bf01-fc6736e84788.png)

Nothing in the vscode console which means no UDP request.

omarcostahamido commented:
@KaneBetter well it seems that you haven't updated `pvm.py` in response to https://github.com/omarcostahamido/PVM/pull/38/commits/33404ff646efb2093db36b51a4a6df5158cb5c23 which I [did](https://github.com/omarcostahamido/PVM/pull/38#issuecomment-1312804162) per your [request](https://github.com/omarcostahamido/PVM/pull/38#discussion_r1020951438)

omarcostahamido commented on pvm.py:
```diff
@@ -36,29 +37,90 @@ def _init_logger():
+
+'''
+UDP command example:
+/PVM 12:13:14
```
```suggestion

```

KaneBetter replied:
Done

omarcostahamido commented on pvm.py:
```diff
@@ -36,29 +37,90 @@ def _init_logger():
+'''
+UDP command example:
+/PVM 12:13:14
+/PVM <action_name> <action_value>
```
```suggestion
/PVM HH MM SS <action_name> <action_value>
```
I could also suggest changing `<action_name>` to `<action_command>` or simply making it `<command> <value>`

KaneBetter replied:
Done

omarcostahamido commented on pvm.py:
```diff
@@ -36,29 +37,90 @@ def _init_logger():
 	global IS_FILE_SET
+	global NEXT_TIME
+	# Parse the time send in UDP, 
+	# From /PVM 12 13 14 to 12:13:14
+	if len(args) == 4:
```
this will not happen now, remember the message format is `/PVM HH MM SS <action_name> <action_value>`
you need to slice the first arguments here first and then pass the rest to be parsed below

KaneBetter replied:
Done

omarcostahamido commented on pvm.py:
```diff
@@ -36,29 +37,90 @@ def _init_logger():
+		return
+	
+	# UDP Command example: /PVM <action_name> <action_value>
 	command = args[1]
```
as is it right now, `args[1]` == `HH`, remember that the format now is `/PVM HH MM SS <action_name> <action_value>`

KaneBetter replied:
Done

omarcostahamido reviewed (changes requested):
please see below, the reason why `pvm.py` is seemingly not received the UDP messages.

KaneBetter commented:
> 33404ff

Yes, I didn't update pvm.py. I just try to log all args in line 44 in the picture. Still, nothing was received. But with the version in the main branch, it works well.

KaneBetter commented:
Main branch
![image](https://user-images.githubusercontent.com/37838032/202042612-022cd1eb-0449-438f-996d-23b26f5da564.png)

KaneBetter commented:
AddTimeInUDP Branch
![image](https://user-images.githubusercontent.com/37838032/202042849-1c77075f-48cb-4a5c-acfc-b78faa7aa471.png)
Which should log out `'PVM', 'HH', 'MM', 'SS', 'file', 'jellyfish720.mp4'`

omarcostahamido commented:
@KaneBetter ok I found the problem. Give me a min, I'll push commit to solve it.

omarcostahamido committed:
fix UDP message format in `pvm_send` abstraction

- now it should output `/PVM HH MM SS <command> <value>`

omarcostahamido commented:
@KaneBetter please fetch this new commit. It should be sending the proper message now. Thx

KaneBetter committed:
Change time parse logic

KaneBetter committed:
Merge branch 'AddTimeInUDP' of https://github.com/omarcostahamido/PVM into AddTimeInUDP

KaneBetter commented:
Changed time parse logic and also increase time accuracy to 0.005 seconds.

omarcostahamido commented on pvm.py:
```diff
@@ -37,18 +38,48 @@ def _init_logger():
+	next_time = datetime.strptime(time_str, "%H:%M:%S") + timedelta(seconds=3)
+
+	# Get command
+	command= args[4]
```
```suggestion
	command = args[4]
```

KaneBetter replied:
Done

omarcostahamido commented on pvm.py:
```diff
@@ -37,18 +38,48 @@ def _init_logger():
+	if len(args) == 6:
+		value = args[5]
+		message += ", value: " + value
+	_logger.info("Start execute %s", message)
```
I guess it would make more sense to print the log entry after the command is successfully executed, or something else if it fails. but I'm not going to make that halt your development strat here.

KaneBetter replied:
Agreed, remove redundant log

omarcostahamido reviewed (approved):
i've noted 2 things below, but I am approving it already.
Looking forward to hear about how it behaves irl, and trying it out with multiple rpi devices at some point.

KaneBetter committed:
Remove redundant log
